doc/_gallery/plot_2_7_multicomp_nonstat_stft.py

```diff
@@ -19,48 +19,88 @@
-axScatter.set_xlabel('Time')
-divider = make_axes_locatable(axScatter)
-axTime = divider.append_axes("top", 1.2, pad=0.5)
-axFreq = divider.append_axes("left", 1.2, pad=0.5)
-axTime.plot(np.real(x))
-axTime.set_xticklabels([])
-axTime.set_xlim(0, N)
-axTime.set_ylabel('Real part')
-axTime.set_title('Signal in time')
-axTime.grid(True)
-axFreq.plot((abs(np.fft.fftshift(np.fft.fft(x))) ** 2)[::-1][:64], f[:64])
-axFreq.set_yticklabels([])
-axFreq.set_xticklabels([])
-axFreq.invert_xaxis()
-axFreq.set_ylabel('Spectrum')
-axFreq.grid(True)
-plt.show()
+
+def generate_signal(nr_samples):
+    x1, _ = fmlin(nr_samples, 0, 0.2)
+    x2, _ = fmlin(nr_samples, 0.3, 0.5)
+    return x1 + x2
+
+
+def plot_stft_contours(axis, ts, freqs, tfr):
+    t_mesh, f_mesh = np.meshgrid(ts, freqs)
+    nyquist = freqs.shape[0] // 2
+    axis.contour(t_mesh[:nyquist, :], f_mesh[:nyquist], tfr[:nyquist, :], 5)
+    axis.grid(True)
+    axis.set_title('Squared modulus of STFT')
+    axis.set_ylabel('Frequency')
+    axis.yaxis.set_label_position("right")
+    axis.set_xlabel('Time')
+    return axis
+
+
+def plot_signal(axis, x, nr_samples):
+    axis.plot(np.real(x))
+    axis.set_xticklabels([])
+    axis.set_xlim(0, nr_samples)
+    axis.set_ylabel('Real part')
+    axis.set_title('Signal in time')
+    axis.grid(True)
+
+
+def plot_spectrum(axis, squared_modulus_spectrum, freqs, nyquist):
+    len_freqs = squared_modulus_spectrum.shape[0] // 2
+    f = np.linspace(0.0, 0.5, len_freqs, endpoint=True)
+    axis.plot(squared_modulus_spectrum[len_freqs:], f)
+    axis.set_yticklabels([])
+    axis.set_xticklabels([])
+    axis.invert_xaxis()
+    axis.set_ylabel('Spectrum')
+    axis.grid(True)
+
+
+def plot(nr_samples, x, tfr, ts, freqs, squared_modulus_spectrum, nyquist):
+    fig, axScatter = plt.subplots(figsize=(10, 8))
+
+    plot_stft_contours(axScatter, ts, freqs, tfr)
+
+    divider = make_axes_locatable(axScatter)
+    axTime = divider.append_axes("top", 1.2, pad=0.5)
+    axFreq = divider.append_axes("left", 1.2, pad=0.5)
+
+    plot_signal(axTime, x, nr_samples)
+    plot_spectrum(axFreq, squared_modulus_spectrum, freqs, nyquist)
+    plt.show()
+
+
+def main():
+    nr_samples = 128
+    x = generate_signal(nr_samples)
+    n_fbins = nr_samples
+    nperseg = 33
+    noverlap = nperseg - 1
+    window = hamming(nperseg)
+    nfft = 128
+    nyquist = nfft // 2
+    stft = ShortTimeFourierTransform(x, timestamps=None, n_fbins=n_fbins)
+    tfr, ts, freqs = stft.run(
+        nfft=nfft,
+        nperseg=nperseg,
+        noverlap=noverlap,
+        return_onesided=False,
+        window=window,
+        scaling="psd")
+
+    threshold = np.amax(np.abs(tfr)) * 0.05
+
+    tfr[np.abs(tfr) <= threshold] = 0.0 + 1j * 0.0
+    tfr = np.abs(tfr) ** 2
+    squared_modulus_spectrum = abs(np.fft.fftshift(np.fft.fft(x))) ** 2
+    plot(nr_samples, x, tfr, ts, freqs, squared_modulus_spectrum, nyquist)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
This change is not necessary to the current issue. Please revert.

Reply:
However, it seems like this script is doing a lot of computation before plotting. That should be the job of the `.plot()` method of the STFT class - not the responsibility of this script. I'll take a look at it later.

Reply:
Ok, I will revert these changes.

Note that you will get the following warning when running the file.
UserWarning: nperseg = 256 is greater than input length  = 128, using nperseg = 128
  warnings.warn('nperseg = {0:d} is greater than input length '

I think it would be better to set nperseg to the correct value than letting Scipy override it.